File: backend/app/HTTP/controllers/data controller.php
<?php

namespace App\Http\Controllers;

use Illuminate\Http\Request;
use Illuminate\Http\JsonResponse;

class DataController extends Controller
{
    /**
     * Get all data sources
     */
    public function getDataSources(): JsonResponse
    {
        $dataSources = [
            [
                'id' => 'ecommerce',
                'name' => 'E-commerce Transactions',
                'description' => 'Customer purchase data with transaction details',
                'enabled' => true,
                'recordCount' => 15000,
                'schema' => 'ecommerce.transactions',
                'lastUpdated' => now()->subHours(2)->toISOString()
            ],
            [
                'id' => 'iot',
                'name' => 'IoT Sensor Data',
                'description' => 'Temperature, Humidity, Pressure readings',
                'enabled' => true,
                'recordCount' => 8500,
                'schema' => 'iot.sensor_data',
                'lastUpdated' => now()->subMinutes(15)->toISOString()
            ],
            [
                'id' => 'logs',
                'name' => 'Web Server Logs',
                'description' => 'Apache/Nginx access and error logs',
                'enabled' => false,
                'recordCount' => 12000,
                'schema' => 'analytics.web_logs',
                'lastUpdated' => now()->subHours(1)->toISOString()
            ],
            [
                'id' => 'financial',
                'name' => 'Financial Market Data',
                'description' => 'Stock prices, trading volumes, market indices',
                'enabled' => false,
                'recordCount' => 6500,
                'schema' => 'finance.market_data',
                'lastUpdated' => now()->subMinutes(30)->toISOString()
            ],
            [
                'id' => 'social',
                'name' => 'Social Media Sentiment',
                'description' => 'Twitter/Facebook posts with sentiment analysis',
                'enabled' => false,
                'recordCount' => 4200,
                'schema' => 'social.sentiment_data',
                'lastUpdated' => now()->subHours(3)->toISOString()
            ]
        ];

        return response()->json($dataSources);
    }

    /**
     * Update data source configuration
     */
    public function updateDataSource(string $id, Request $request): JsonResponse
    {
        $request->validate([
            'enabled' => 'boolean'
        ]);

        // In a real implementation, this would update the database
        return response()->json([
            'id' => $id,
            'enabled' => $request->enabled,
            'updated_at' => now()->toISOString(),
            'message' => 'Data source updated successfully'
        ]);
    }

    /**
     * Get sample data for a data source
     */
    public function getSampleData(string $id): JsonResponse
    {
        $sampleData = $this->generateSampleData($id);
        
        return response()->json([
            'dataSourceId' => $id,
            'sampleSize' => count($sampleData),
            'data' => $sampleData,
            'generatedAt' => now()->toISOString()
        ]);
    }

    /**
     * Generate sample data based on data source type
     */
    private function generateSampleData(string $sourceType): array
    {
        switch ($sourceType) {
            case 'ecommerce':
                return $this->generateEcommerceData();
            case 'iot':
                return $this->generateIoTData();
            case 'logs':
                return $this->generateLogData();
            case 'financial':
                return $this->generateFinancialData();
            case 'social':
                return $this->generateSocialData();
            default:
                return [];
        }
    }

    private function generateEcommerceData(): array
    {
        $products = ['Laptop', 'Phone', 'Tablet', 'Watch', 'Headphones'];
        $data = [];

        for ($i = 0; $i < 10; $i++) {
            $data[] = [
                'id' => $i + 1,
                'product' => $products[array_rand($products)],
                'amount' => rand(50, 1000),
                'timestamp' => now()->subHours(rand(1, 24))->toISOString(),
                'customer_id' => 'cust_' . rand(1000, 9999)
            ];
        }

        return $data;
    }

    private function generateIoTData(): array
    {
        $data = [];

        for ($i = 0; $i < 10; $i++) {
            $data[] = [
                'device_id' => 'sensor_' . rand(1, 20),
                'temperature' => rand(10, 50),
                'humidity' => rand(20, 100),
                'pressure' => rand(950, 1000),
                'timestamp' => now()->subMinutes(rand(1, 60))->toISOString()
            ];
        }

        return $data;
    }

    private function generateLogData(): array
    {
        $ips = ['192.168.1.1', '10.0.0.1', '172.16.0.1'];
        $endpoints = ['/api/users', '/api/orders', '/dashboard'];
        $data = [];

        for ($i = 0; $i < 10; $i++) {
            $data[] = [
                'ip' => $ips[array_rand($ips)],
                'endpoint' => $endpoints[array_rand($endpoints)],
                'status_code' => rand(0, 10) < 9 ? 200 : 404,
                'response_time' => rand(50, 1000),
                'timestamp' => now()->subMinutes(rand(1, 60))->toISOString()
            ];
        }

        return $data;
    }

    private function generateFinancialData(): array
    {
        $symbols = ['AAPL', 'GOOGL', 'MSFT', 'AMZN'];
        $data = [];

        for ($i = 0; $i < 10; $i++) {
            $data[] = [
                'symbol' => $symbols[array_rand($symbols)],
                'price' => rand(50, 350) + rand(0, 99) / 100,
                'volume' => rand(100000, 1000000),
                'change' => (rand(0, 1) ? 1 : -1) * (rand(0, 10) + rand(0, 99) / 100),
                'timestamp' => now()->subMinutes(rand(1, 60))->toISOString()
            ];
        }

        return $data;
    }

    private function generateSocialData(): array
    {
        $platforms = ['Twitter', 'Facebook', 'Instagram'];
        $sentiments = ['positive', 'negative', 'neutral'];
        $data = [];

        for ($i = 0; $i < 10; $i++) {
            $data[] = [
                'platform' => $platforms[array_rand($platforms)],
                'sentiment' => $sentiments[array_rand($sentiments)],
                'score' => rand(0, 100) / 100,
                'mentions' => rand(1, 1000),
                'timestamp' => now()->subMinutes(rand(1, 60))->toISOString()
            ];
        }

        return $data;
    }
}

File: Rails-backend/app/controllers/api/V1/data controller.rb
class Api::V1::DataController < ApplicationController
  before_action :handle_options_request

  # GET /api/v1/data
  def index
    page = params[:page] || 1
    per_page = params[:per_page] || 50
    data_type = params[:type] || 'all'

    # Simulate data retrieval
    data = generate_sample_data(data_type, per_page.to_i)
    
    render json: {
      data: data,
      pagination: {
        current_page: page.to_i,
        per_page: per_page.to_i,
        total_pages: 10,
        total_count: 500
      },
      metadata: {
        data_type: data_type,
        generated_at: Time.current
      }
    }
  end

  # POST /api/v1/data
  def create
    data_params = params.require(:data).permit(:type, :count, :format)
    
    generated_data = case data_params[:type]
    when 'ecommerce'
      generate_ecommerce_data(data_params[:count].to_i || 1000)
    when 'iot'
      generate_iot_data(data_params[:count].to_i || 1000)
    when 'financial'
      generate_financial_data(data_params[:count].to_i || 1000)
    when 'social'
      generate_social_data(data_params[:count].to_i || 1000)
    else
      generate_mixed_data(data_params[:count].to_i || 1000)
    end

    render json: {
      success: true,
      message: 'Data generated successfully',
      count: generated_data.length,
      type: data_params[:type],
      data: generated_data.first(10) # Return first 10 records as preview
    }
  end

  # GET /api/v1/data/sources
  def sources
    sources = [
      {
        id: 'ecommerce',
        name: 'E-commerce Transactions',
        description: '10k+ transaction records',
        status: 'active',
        last_updated: 1.hour.ago,
        record_count: rand(8000..12000)
      },
      {
        id: 'iot',
        name: 'IoT Sensor Data',
        description: 'Temperature, Humidity, Pressure sensors',
        status: 'active',
        last_updated: 5.minutes.ago,
        record_count: rand(15000..25000)
      },
      {
        id: 'logs',
        name: 'Web Server Logs',
        description: 'Apache/Nginx access logs',
        status: 'active',
        last_updated: 2.minutes.ago,
        record_count: rand(50000..100000)
      },
      {
        id: 'financial',
        name: 'Financial Market Data',
        description: 'Stock prices and trading volumes',
        status: 'active',
        last_updated: 30.seconds.ago,
        record_count: rand(5000..8000)
      },
      {
        id: 'social',
        name: 'Social Media Sentiment',
        description: 'Twitter, Facebook sentiment analysis',
        status: 'active',
        last_updated: 1.minute.ago,
        record_count: rand(20000..35000)
      }
    ]

    render json: { sources: sources }
  end

  # POST /api/v1/data/generate
  def generate
    data_type = params[:type]
    count = params[:count].to_i || 1000
    
    case data_type
    when 'ecommerce'
      data = generate_ecommerce_data(count)
    when 'iot'
      data = generate_iot_data(count)
    when 'financial'
      data = generate_financial_data(count)
    when 'social'
      data = generate_social_data(count)
    else
      data = generate_mixed_data(count)
    end

    render json: {
      success: true,
      type: data_type,
      count: data.length,
      sample: data.first(5)
    }
  end

  # GET /api/v1/data/export
  def export
    format = params[:format] || 'json'
    data_type = params[:type] || 'all'
    
    data = generate_sample_data(data_type, 1000)
    
    case format
    when 'csv'
      csv_data = generate_csv(data)
      send_data csv_data, filename: "export_#{data_type}_#{Date.current}.csv", type: 'text/csv'
    when 'json'
      send_data data.to_json, filename: "export_#{data_type}_#{Date.current}.json", type: 'application/json'
    else
      render json: { error: 'Unsupported format' }, status: :bad_request
    end
  end

  # POST /api/v1/data/validate
  def validate
    data = params[:data]
    validation_rules = params[:rules] || {}
    
    results = {
      total_records: data&.length || 0,
      valid_records: 0,
      invalid_records: 0,
      validation_errors: [],
      quality_score: 0
    }

    if data.present?
      results[:valid_records] = (data.length * 0.95).to_i
      results[:invalid_records] = data.length - results[:valid_records]
      results[:quality_score] = (results[:valid_records].to_f / data.length * 100).round(2)
    end

    render json: results
  end

  # GET /api/v1/data/lineage
  def lineage
    lineage_data = {
      nodes: [
        { id: 'sources', label: 'Data Sources', type: 'source', x: 50, y: 50 },
        { id: 'ingestion', label: 'Data Ingestion', type: 'process', x: 200, y: 50 },
        { id: 'raw_storage', label: 'Raw Storage', type: 'storage', x: 350, y: 30 },
        { id: 'processing', label: 'Data Processing', type: 'process', x: 350, y: 90 },
        { id: 'clean_storage', label: 'Clean Storage', type: 'storage', x: 500, y: 30 },
        { id: 'analytics', label: 'Analytics', type: 'process', x: 500, y: 90 },
        { id: 'reports', label: 'Reports', type: 'output', x: 650, y: 60 }
      ],
      links: [
        { source: 'sources', target: 'ingestion' },
        { source: 'ingestion', target: 'raw_storage' },
        { source: 'ingestion', target: 'processing' },
        { source: 'raw_storage', target: 'processing' },
        { source: 'processing', target: 'clean_storage' },
        { source: 'processing', target: 'analytics' },
        { source: 'clean_storage', target: 'analytics' },
        { source: 'analytics', target: 'reports' }
      ]
    }

    render json: lineage_data
  end

  private

  def generate_sample_data(type, count)
    case type
    when 'ecommerce'
      generate_ecommerce_data(count)
    when 'iot'
      generate_iot_data(count)
    when 'financial'
      generate_financial_data(count)
    when 'social'
      generate_social_data(count)
    else
      generate_mixed_data(count)
    end
  end

  def generate_ecommerce_data(count)
    products = ['Laptop', 'Phone', 'Tablet', 'Watch', 'Headphones', 'Camera']
    
    Array.new(count) do |i|
      {
        id: i + 1,
        product: products.sample,
        amount: rand(50..1000),
        quantity: rand(1..5),
        customer_id: rand(1000..9999),
        timestamp: rand(30.days).seconds.ago,
        category: ['Electronics', 'Accessories'].sample
      }
    end
  end

  def generate_iot_data(count)
    Array.new(count) do |i|
      {
        device_id: "sensor_#{rand(1..100)}",
        temperature: rand(10.0..40.0).round(2),
        humidity: rand(20.0..80.0).round(2),
        pressure: rand(950.0..1050.0).round(2),
        timestamp: rand(7.days).seconds.ago,
        location: ['Building A', 'Building B', 'Warehouse'].sample
      }
    end
  end

  def generate_financial_data(count)
    symbols = ['AAPL', 'GOOGL', 'MSFT', 'AMZN', 'TSLA']
    
    Array.new(count) do |i|
      base_price = rand(50.0..300.0)
      {
        symbol: symbols.sample,
        price: base_price.round(2),
        volume: rand(100000..1000000),
        change: (rand(-20.0..20.0)).round(2),
        timestamp: rand(24.hours).seconds.ago
      }
    end
  end

  def generate_social_data(count)
    platforms = ['Twitter', 'Facebook', 'Instagram', 'LinkedIn']
    sentiments = ['positive', 'negative', 'neutral']
    
    Array.new(count) do |i|
      {
        platform: platforms.sample,
        sentiment: sentiments.sample,
        score: rand(0.0..1.0).round(3),
        mentions: rand(1..1000),
        hashtags: rand(0..5),
        timestamp: rand(24.hours).seconds.ago
      }
    end
  end

  def generate_mixed_data(count)
    types = ['ecommerce', 'iot', 'financial', 'social']
    data = []
    
    types.each do |type|
      data += send("generate_#{type}_data", count / 4)
    end
    
    data.shuffle.first(count)
  end

  def generate_csv(data)
    return '' if data.empty?
    
    headers = data.first.keys
    CSV.generate do |csv|
      csv << headers
      data.each { |row| csv << headers.map { |h| row[h] } }
    end
  end
end